### ISSUE TITLE
Performance problems on large, multi-package monorepos

### ISSUE THREAD

**Issue by drew-walker** [jest-community/vscode-jest#1030]
### Environment

1. `vscode-jest version`: 5.2.3
2. `node -v`: 16.17.0
3. `npm -v` or `yarn --version`: 8.15.0
4. `npm ls jest` or `npm ls react-scripts` (if you haven’t ejected): 29.3.1 is the only version
5. your vscode-jest settings if customized:
    - jest.jestCommandLine? default
    - jest.autoRun? `{ "watch": false, "onSave": "test-src-file" }`
    - anything else that you think might be relevant? Nope
6. Operating system: Ventura 13.3.1

### Prerequisite
- are you able to run jest test from the command line? Yes
- how do you run your tests from the command line? `npm test`

### Steps to Reproduce

In a multi-package mono-repo with many packages (we currently have around 120), the extension is running jest with `--listTests --json --watchAll=false` for each package on startup to populate the test explorer. This isn't scalable from a memory usage perspective, and is now creating so much memory pressure on startup that it takes around 1 minute on an M1 for VSCode to become usable. I've checked out the source and I see no way of avoiding this behavior (currently).

I'd like to propose a new setting to address the problem. The setting would be called `updateTestFileList` with values of `on-start` or `on-test-run` (with a default value of `on-start` to match existing behavior). With the `on-test-run` setting, the `updateTestFileList` function would not be called on startup, and would instead only run (for the relevant package) when running a test/tests. The implications of the change are that the test explorer would not be populated on startup, but only when running tests. I personally believe this is a good experience trade-off for huge repos. I'm happy to create a pull request for this change, but I wanted to discuss it first.

1. Is this a palatable solution for you?
2. How do you feeling about the naming suggestions I've made?
3. Assuming you're ok with my suggestions, is there anything you'd like to change about the solution before I create a PR?

### Expected Behavior

VSCode should be usable to edit files very soon after opening or reloading.

### Actual Behavior

Memory usage spikes to the point where memory pressure is so high that it's impossible to type/scroll/open files in VSCode.

**Comment by connectdotz** [comment]
Indeed, this is an interesting use case. 2 things come to mind:
1. If use a separate option like `updateTestFileList`, it might conflict with autoRun, or people would have to adjust 2 options to achieve the correct effect. I wonder if we should consider expanding autoRun instead.
2. As you realized, without find out the test file list, some UI will not be available: `The implications of the change are that the test explorer would not be populated on startup, but only when running tests.` The question is how will one run tests in this situation? via command palette (we already have a start runner command)? via context menu (there might be some issue without knowing if the file is a test file)? Or something else you have in mind? Keep in mind that the gutter status icon is managed by TestExplorer, it will not be there if we didn't know what the test files are and publish the test items accordingly...

**Comment by drew-walker** [comment]
Thanks for responding. Here are my thoughts:

> If use a separate option like updateTestFileList, it might conflict with autoRun, or people would have to adjust 2 options to achieve the correct effect. I wonder if we should consider expanding autoRun instead.

Yeah, I see what you mean here. Especially when `autoRun` is set to `watch`. It would become a contradictory setting to `updateTestFileList: "on-test-run"`. I was considering instead using `jest.autoRun.listTests: "on-startup" | "on-test-run"`, but that would still contradict `jest.autoRun: { watch: true, onStartup: ["all-tests" }`. I feel like this might be because of the fact that we actually want to contradict the default behavior. Perhaps an answer here could be to allow the contradictory settings to exist, but present a message when the settings are contradictory. e.g.

_Using `autoRun: "watch"` requires test information to be to collected on startup. Either switch to `listTests: "on-startup"`, or turn of auto-run. Turning off auto-run can improve startup performance in projects with large test suites._

Another alternative could also be to disable auto-run by default, and instead make it an opt-in setting. But that would be a breaking change for people that upgrade.

> As you realized, without find out the test file list, some UI will not be available: The implications of the change are that the test explorer would not be populated on startup, but only when running tests. The question is how will one run tests in this situation? via command palette (we already have a start runner command)? via context menu (there might be some issue without knowing if the file is a test file)? Or something else you have in mind? Keep in mind that the gutter status icon is managed by TestExplorer, it will not be there if we didn't know what the test files are and publish the test items accordingly...

From my local testing, the behavior with the setting set to "on-test-run" seems quite intuitive:
1. Packages are still listed in the Test Explorer and all tests for a package can be invoked with the "Run Test" button. The individual tests in that package won't be visible until you use "Run Test", but will be for the remainder of your VSCode session.
2. I actually didn't realize tests could be invoked from the command palette until now, but this seems to continue to work as expected. The tests are discovered an ran during the test run after the command is invoked.
3. The gutter status icon doesn't seem to be a problem. I'm not sure what's triggering this behavior yet, but when I visit a test it is automatically added to the Test Explorer. This means I can visit any test and continue to use the gutter to run the suite or an individual test.

Would you generally say you're open to this type of change at the moment? If so, I can also arrange a pull request while we're continuing the discussion.

**Comment by drew-walker** [comment]
I guess a completely different option would be to create a limit on the number of jest processes that can be spawned in parallel. e.g. Limit it to the number of cores - 1 (or provide a setting). That will avoid exhausting all memory with large numbers of packages.

**Comment by connectdotz** [comment]
> I guess a completely different option would be to create a limit on the number of jest processes that can be spawned in parallel.

Limiting the number of concurrent processes felt arbitrary and might still be insufficient for some use cases. I like the idea to have a setting to control each folder explicitly.

There is a new setting `enable` (see it in README.md in master) that was planned for the upcoming v6 release. I am thinking maybe we could expand it beyond the current boolean value, such as `true|false|'deferred'`. Since this setting sits on top of autoRun (and everything else), it is clear that it will supersede the autoRun config. It could do exactly the same as your current implementation.

Regarding UI - yes, initially only the package item will be visible in TestExplorer. As you open test files in the editors the extension will parse the code and publish the test block identified. However, without knowing which file is a test file, it will try to parse every code file, so you might see some warning/error message for parsing errors because they are not test files. If we disable this, then there will be no gutter indicator in the "deferred" extension. Fortunately, I think we did cache the parse error, so it should not repeatedly parse-and-fail, until the code is changed/saved. Anyway, we might need to play with it to make sure it is not a problem in real-world use cases.

I am definitely open to a PR. But advise waiting a little bit as there is some significant code change coming for v6, mainly related to virtual folders. It will make code merge much easier if you can wait until that baseline gets merged into the master, which I plan to do this weekend.

**Comment by connectdotz** [comment]
@drew-walker FYI, the main v6 code change is in, feel free to submit the PR you were working on.

**Comment by albertogasparin** [comment]
Hi! Any progress on this front? I'm also working on a very large codebase and at every boot or large merge/rebase there are hundreds node processes spawning and completely freezing the UI. 
I love the ability to run tests on save, but currently I have to disable it every time that I'm not actually testing as it makes VSCode very unstable.

**Comment by drew-walker** [comment]
I haven't had a chance to get back to this yet. It's definitely still a problem in our repo, so I'll try and set aside some time to progress my PR in the next couple of weeks.

**Comment by connectdotz** [comment]
There is a new proposal related to this topic: #1058. Please feel free to chime in your thoughts about it.
